ccl/changefeedccl: TestChangefeedPanicRecovery failed

ccl/changefeedccl.TestChangefeedPanicRecovery [failed](https://tanzanite.cluster.engflow.com/invocations/default/e93cd932-393b-4db7-ac96-fe7fe42019e3?testReportRun=22&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-24.1.0-rc @ [5e4ca9e26f1a25681de9c944298cfa139c344466](https://github.com/cockroachdb/cockroach/commits/5e4ca9e26f1a25681de9c944298cfa139c344466):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestChangefeedPanicRecovery (14m30s)
	TestChangefeedPanicRecovery/cloudstorage#01 (14m28s)

```
Stack: 

```
goroutine 842916 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2366 +0x385
created by time.goFunc
	GOROOT/src/time/sleep.go:177 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestChangefeedPanicRecovery
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestChangefeedPanicRecovery2444363071
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestChangefeedPanicRecovery/cloudstorage
    helpers_test.go:867: making server as system tenant
    helpers_test.go:943: making cloudstorage feed factory
    testfeed_test.go:283: creating external connection
    testfeed_test.go:286: ran create external connection
=== RUN   TestChangefeedPanicRecovery/cloudstorage#01
    helpers_test.go:864: making server as secondary tenant
    helpers_test.go:943: making cloudstorage feed factory
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=22</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedPanicRecovery.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39101

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -5681,6 +5681,19 @@ func TestChangefeedPanicRecovery(t *testing.T) {
 		sqlDB.Exec(t, `INSERT INTO foo(id, s) VALUES (0, 'hello'), (1, null)`)
 	}
 
+	waitForFeedErr := func(t *testing.T, feed cdctest.TestFeed, timeout time.Duration) error {
+		start := timeutil.Now()
+		for {
+			if time.Since(start) >= timeout {
+				t.Fatalf("feed did not return error before timeout of %s", timeout)
+			}
+			_, err := feed.Next()
+			if err != nil {
+				return err
+			}
+		}
+	}
+
 	cdcTest(t, func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 		prep(t, sqlDB)
@@ -5691,32 +5704,28 @@ func TestChangefeedPanicRecovery(t *testing.T) {
 	})
 
 	// Check that all panics while evaluating the WHERE clause in an expression are recovered from.
+	// NB: REPAIRCLUSTER is required to use crdb_internal.force_panic.
 	cdcTest(t, func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 		prep(t, sqlDB)
 		foo := feed(t, f,
 			`CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT 1 FROM foo WHERE crdb_internal.force_panic('wat') IS NULL`)
 		defer closeFeed(t, foo)
-		var err error
-		for err == nil {
-			_, err = foo.Next()
-		}
-		require.Error(t, err, "error while evaluating WHERE clause")
-	})
+		err := waitForFeedErr(t, foo, 2*time.Minute)
+		require.ErrorContains(t, err, "error evaluating CDC expression", "expected panic recovery while evaluating WHERE clause")
+	}, feedTestAdditionalSystemPrivs("REPAIRCLUSTER"))
 
 	// Check that all panics while evaluating the SELECT clause in an expression are recovered from.
+	// NB: REPAIRCLUSTER is required to use crdb_internal.force_panic.
 	cdcTest(t, func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 		prep(t, sqlDB)
 		foo := feed(t, f,
 			`CREATE CHANGEFEED WITH schema_change_policy='stop' AS SELECT crdb_internal.force_panic('wat') FROM foo`)
 		defer closeFeed(t, foo)
-		var err error
-		for err == nil {
-			_, err = foo.Next()
-		}
-		require.Error(t, err, "error while evaluating SELECT clause")
-	})
+		err := waitForFeedErr(t, foo, 2*time.Minute)
+		require.ErrorContains(t, err, "error evaluating CDC expression", "expected panic recovery while evaluating SELECT clause")
+	}, feedTestAdditionalSystemPrivs("REPAIRCLUSTER"))
 }
 
 func TestChangefeedPauseUnpause(t *testing.T) {

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -551,6 +551,7 @@ type feedTestOptions struct {
 	allowedSinkTypes             []string
 	disabledSinkTypes            []string
 	settings                     *cluster.Settings
+	additionalSystemPrivs        []string
 }
 
 type feedTestOption func(opts *feedTestOptions)
@@ -584,6 +585,12 @@ var feedTestEnterpriseSinks = func(opts *feedTestOptions) {
 
 var feedTestOmitSinks = func(sinkTypes ...string) feedTestOption {
 	return func(opts *feedTestOptions) { opts.disabledSinkTypes = append(opts.disabledSinkTypes, sinkTypes...) }
+}
+
+var feedTestAdditionalSystemPrivs = func(privs ...string) feedTestOption {
+	return func(opts *feedTestOptions) {
+		opts.additionalSystemPrivs = append(opts.additionalSystemPrivs, privs...)
+	}
 }
 
 func (opts feedTestOptions) omitSinks(sinks ...string) feedTestOptions {
@@ -1017,7 +1024,7 @@ func makeFeedFactoryWithOptions(
 				User:   url.UserPassword(u, pass[0]),
 				Host:   s.SQLAddr(), Path: "d"}, func() {}
 		}
-		sink, cleanup := getInitialSinkForSinklessFactory(t, db, pgURLForUserSinkless)
+		sink, cleanup := getInitialSinkForSinklessFactory(t, db, pgURLForUserSinkless, options)
 		root, cleanupRoot := pgURLForUserSinkless(username.RootUser)
 		f := makeSinklessFeedFactory(t, s, sink, root, pgURLForUserSinkless)
 		return f, func() {
@@ -1042,6 +1049,7 @@ func getInitialDBForEnterpriseFactory(
 		user := "EnterpriseFeedUser"
 		password := "hunter2"
 		createUserWithDefaultPrivilege(t, rootDB, user, password, "CHANGEFEED", "SELECT")
+		grantUserAdditionalSystemPrivileges(t, rootDB, user, opts.additionalSystemPrivs)
 		pgURL := url.URL{
 			Scheme: "postgres",
 			User:   url.UserPassword(user, password),
@@ -1058,16 +1066,17 @@ func getInitialDBForEnterpriseFactory(
 }
 
 func getInitialSinkForSinklessFactory(
-	t *testing.T, db *gosql.DB, sinkForUser sinkForUser,
+	t *testing.T, db *gosql.DB, sinkForUser sinkForUser, opts feedTestOptions,
 ) (url.URL, func()) {
 	// Instead of creating sinkless changefeeds on the root connection, we may choose to create
 	// them on a test user connection. This user should have the minimum privileges to create a changefeed,
 	// which means they default to having the SELECT privilege on all tables.
 	const percentNonRoot = 1
-	if rand.Float32() < percentNonRoot {
+	if !opts.forceRootUserConnection && rand.Float32() < percentNonRoot {
 		user := "SinklessFeedUser"
 		password := "hunter2"
 		createUserWithDefaultPrivilege(t, db, user, password, "SELECT")
+		grantUserAdditionalSystemPrivileges(t, db, user, opts.additionalSystemPrivs)
 		return sinkForUser(user, password)
 	}
 	return sinkForUser(username.RootUser)
@@ -1093,6 +1102,15 @@ func createUserWithDefaultPrivilege(
 		if err != nil {
 			t.Fatal(err)
 		}
+	}
+}
+
+func grantUserAdditionalSystemPrivileges(
+	t *testing.T, rootDB *gosql.DB, user string, privs []string,
+) {
+	for _, priv := range privs {
+		_, err := rootDB.Exec(fmt.Sprintf(`GRANT SYSTEM %s TO %s`, priv, user))
+		require.NoError(t, err)
 	}
 }
 

